### ISSUE TITLE
commercial DRM test coverage

### ISSUE THREAD

**Issue by jpiesing** [cta-wave/device-playback-task-force#118]
This is my attempt to capture ideas on coverage of commercial DRM tests if WAVE would do these in 2024. I think there are several groups of possible tests.

1. Repetitions of existing tests but with encrypted content 
(Each of these would be repeated for each DRM system).
- 8.2 Sequential Track Playback and/or 8.12 Playback of Encrypted Content and/or 9.2 Regular Playback of a CMAF Presentation
- 8.3 Random Access to Fragment and 8.4 Random Access to Time and/or 9.3 Random Access of a WAVE Presentation)
- 8.5 Switching Set Playback (but perhaps with fewer Representations)
- 8.13 Restricted Splicing of Encrypted Content where the encrypted content is DRM protected
- 8.14 Sequential Playback of Encrypted and Non-Encrypted Baseline Content where the encrypted content is DRM protected (perhaps with a variation where this happens 5-10 times)

2. General DRM features
(Each of these would be repeated for each DRM system as far as applicable)
- Key changes before the end of the content - there may be multiple variations on this, e.g. end of one content item and start of the next and/or a time-limited license expires
- Key change and a new license is obtained once
- Key change and a new license is obtained (both repeated at least 5-10 times)
- Key rotation (if different from the previous line)
- The EME update method to the extent it's not included in the key change tests above.
- Persistent (pre-loaded) licenses
- Encrypted video with unencrypted audio
- The EME Get Supported Configuration algorithm
- Multiple EME initialization segments (i.e. pssh boxes)
3. DRM system specific features
3.1 PlayReady
- SL3000 for video (assuming repetitions of existing tests use SL2000)
- SL3000 for video and SL2000 for audio (repeated for both cenc and cbcs)
- MediaKeySystemMediaCapability.robustness
- Custom Data
- License server over-ride
3.2 Widevine
- L1 for video (assuming repetitions of existing tests uses L3)
- L1 for video and L3 for audio
- to be completed
3.3 FairPlay
- to be completed
4 Errors
- License does not have the correct rights for the content
- Content is DRM protected but not by one of the supported ones on this implementation
- Content is protected by a supported DRM but no license is available
- License is available but has expired (depending on the DRM system, there may be more than one variation of 'expired', e.g. expires at an absolute (wall-clock) time+date and/or expires at a time relative to when the license was acquired
- License expires during playback (where expires may be an absolute time and/or a relative time as before)
- Video and audio have different KIDs and there's only a license for one of them

**Comment by dsilhavy** [comment]
Not sure if this is covered explicitly: The transition from unencrypted content to encrypted content is something that can cause problems. From my understanding, the MSE/EME stack needs to be initialized with an encrypted init segment first. Otherwise, the MSE needs to be reset prior to the transition.

**Comment by louaybassbouss** [comment]
> Not sure if this is covered explicitly: The transition from unencrypted content to encrypted content is something that can cause problems. From my understanding, the MSE/EME stack needs to be initialized with an encrypted init segment first. Otherwise, the MSE needs to be reset prior to the transition.

@dsilhavy this is covered in the `[8.14 sequential-playback-of-encrypted-and-non-encrypted-baseline-content](https://github.com/cta-wave/dpctf-tests?tab=readme-ov-file#templates)` for use case: Main Content is encrypted, Ad Content unencrypted. There is a transition from encrypted to unencrypted and then another transition from unencrypted to encrypted.

**Comment by mbergman42** [comment]
Question, if we are to be supporting FairPlay/Widevine/Playready on behalf of industry, what support can we expect from the owners of those technologies? Can we get Google/MSFT/Apple SMEs involved, can they help cover development costs?

**Comment by jpiesing** [comment]
Here are a few comments from the discussion in the Fall Forum meeting.
1. What would we be testing & (perhaps more important, what would we not be testing). We would not be testing the implementation of EME or of any particular DRM system but the integration of the browser (i.e. the EME implementation) with the DRM system and the media playback feature of the host OS.
2. A de-facto common set of unit tests for integration of commercial DRM benefits the device manufacturer / implementer / integrator as well as content and app providers - although in the latter case, it's of less benefit to top tier app providers who are big enough to make & maintain their own unit tests for devices and require devices pass those tests as a condition for access to content.
3. The reality on DRM systems is that almost all media consumption devices (outside the Apple Ecosystem) support PlayReady and Widevine - although these are not always available to all applications. For example, a device may support Widevine only in the context of YouTube and not make it available via it's default browser for HTML5 apps. Companies that 10 years ago had their own DRM or pseudo-DRM may now also offer multi-DRM license serving. For example Intertrust offer ExpressPlay as a multi-DRM license serving product and Nagra offer "Multi-DRM" as a multi-DRM licensing serving product. If you look at https://www.ezdrm.com/new-compare-drms, the only additional DRM is the Huawei WisePlay - presumably a consequence of the US sanctions against them.
4. For a list of multi DRM service providers, see https://docs.unified-streaming.com/documentation/drm/drm-providers.html. A new entrant's barrier to entry is supporting a reasonable number of that list to support them. I may try and go through that list to list which DRMs are supported by each one. 
5.

**Comment by jpiesing** [comment]
> Question, if we are to be supporting FairPlay/Widevine/Playready on behalf of industry, what support can we expect from the owners of those technologies? Can we get Google/MSFT/Apple SMEs involved, can they help cover development costs?

I think we can expect very little from the incumbent suppliers.

**Comment by jpiesing** [comment]
One of the discussion points in the fall forum meeting was about the choice and variety of DRM systems. 
One way to get information on this without spending lots of money is to look at the multi-DRM vendors and what DRM systems they support. The attached spreadsheet contains a list of the multi-DRM vendors I could find and what I could find of the DRM systems they support. It contains links to the information used.

It is quite dramatic.
- The lists I found identified 19 multi-DRM service providers
- 3 may not exist any more or may exist under different names or may be duplicates due to M&A activity (Cisco/Synamedia, Conax part of Nagra, Vualto part of JW Player)
- One looks more like a streaming video app than a provider of services to such apps
- Everyone where I could find information supports PlayReady, Widevine and FairPlay.
- Excluding those 3, only 4 other DRMs were mentioned.
- 3 are part of groups / companies that have also have their own DRM and support that DRM (Intertrust ExpressPlay/Marlin, Nagra PRM, ViaAccess Orca DRM). These 3 DRMs are not supported by any other multi-DRM vendor.
- Huawei WisePlay was mentioned by 2 multi-DRM providers, this was the only DRM system other than the 3 incumbents supported by any multi-DRM vendor without some form of common ownership.

I'm happy to be corrected. Improvements welcome.

[Analysis of DRM systems supported by Multi-DRM service providers 2024-10-04 (1).xlsx](https://github.com/user-attachments/files/17256965/Analysis.of.DRM.systems.supported.by.Multi-DRM.service.providers.2024-10-04.1.xlsx)

IMHO this would justify any potential WAVE activity in this field focussing on the 3 DRM systems that are very widely supported.

**Comment by haudiobe** [comment]
TF Call:
- thank you - great info
- we have also some material in the survey @cta-source will share
- We keep [this excel](https://github.com/user-attachments/files/17256965/Analysis.of.DRM.systems.supported.by.Multi-DRM.service.providers.2024-10-04.1.xlsx) as record to justify that we focus on only the 3 DRM systems Fairplay, PlayReady and Widevine

**Comment by gitwjr** [comment]
Hi Thomas,

Please send the ppt from today’s call. I didn’t receive it if you already sent it out.

Thanks,
Bill

**Comment by gitwjr** [comment]
Never mind. I found it in my “other” folder.

Bill


From: Bill Rose
Sent: Wednesday, October 9, 2024 12:06 PM
To: cta-wave/device-playback-task-force ***@***.***>
Subject: RE: [cta-wave/device-playback-task-force] commercial DRM test coverage (Issue #118)

Hi Thomas,

Please send the ppt from today’s call. I didn’t receive it if you already sent it out.

Thanks,
Bill

**Comment by jpiesing** [comment]
Here is my attempt to identify what might be an inexpensive minimum step in this direction. I use EZDRM as an example.

1. Identify which existing tests and test content to clone for the first set of commercial DRM tests. It could be one, it could be all the ClearKey tests, it could be the list in my initial comment above. It could be something else. In order to simplify and keep costs down for the 1st step, I assume it's limited to 1) tests that only need one license request which is at the start and 2) tests which use one of the existing observations so no changes to the OF.
2. Modify the existing content packaging scripts to encrypt test content using commercial DRM. For EZDRM, this is a call to a web API to obtain the key(s) and then using that to encrypt the content. We already have the encryption piece for the ClearKey test content but not the web API call.
3. Modify dpctf-testharness.js to be able to request a key from EZDRM. This may mean modifying tests.json or test-config.json to include information needed for the key request, which may be information from the API call when the content was encrypted. All of this should be done in a way that makes it easy to add other multi-DRM service providers in the future. 
4. Clone the existing HTML files under appropriate names for DRM encrypted content.
5. Create the actual tests by combining the cloned HTML files with the encrypted content
6. Test, test, test
7. Validate

**Comment by jpiesing** [comment]
We have the following tests for ClearKey encrypted content at the moment (only the locally hosted versions selected).

AVC video
* cfhd_14.985_29.97_59.94-local/playback-of-encrypted-content-https__t1-cenc.html
* cfhd_12.5_25_50-local/restricted-splicing-of-encrypted-content-https__splice_main-cenc_splice_ad-cenc.html
* cfhd_12.5_25_50-local/sequential-playback-of-encrypted-and-non-encrypted-baseline-content-https__splice_main-cenc_splice_ad.html
* cfhd_12.5_25_50-local/playback-of-encrypted-content-https__t1-cenc.html
* cfhd_15_30_60-local/playback-of-encrypted-content-https__t1-cenc.html

Audio
* caac_sets-local/sequential-playback-of-encrypted-and-non-encrypted-baseline-content-https__at15_at14.html
* caac_sets-local/restricted-splicing-of-encrypted-content-https__at15_at16.html
* caac_sets-local/playback-of-encrypted-content-https__at5.html

Looking at this, a simple starting point would be the AVC stream t1 in either the 25 or 30Hz variations. This could be used / cloned to produce the following streams;
* PlayReady SL2000
* PlayReady SL3000
* Widevine L1
* Widevine L3
* FairPlay
* PlayReady SL2000+Widevine (tbc L1 / L3)
* PlayReady SL2000+Widevine (tbc L1 / L3) + FairPlay

Each of these would be combined with modified versions of playback-of-encrypted-content-https__t1-cenc and dpctf-testharness.js to generate the same number of tests.

Obviously this can be reduced still further by deferring PlayReady SL3000 and Widevine L1 to later.

A slightly larger starting point would add clones of a few more of the tests listed in the first comment in this issue and encrypted audio.

**Comment by jpiesing** [comment]
Here is a TP Vision contribution to HbbTV that lists things to be tested which is also relevant here. Some of these may duplicate the above.

*  Changing of initialization segments over time including changing between initialization segments that indicate encrypted media and ones that indicate non-encrypted media and vice-versa (see EME [] clause 7.5.2 "Initialization Data Encountered").
* Changing of KIDs
* Expiration of licenses (see MediaKeySession.expiration in EME[]).
* Where a DRM system supports persistent licenses, requesting these using a MediaKeys object with a MediaKeySessionType of "persistent-license" (see EME [] clause 5). The CDM shall not request a new license if it already has one that can be used to decrypt content to be played.
* Providing messages (including licenses) to the CDM ( see MediaKeySession.update() in EME[]).
* Requesting a supported key system configuration (see the Get Supported Configuration algorithm, clause 3.2.2.1 in EME []).

These test that various requirements of EME work however the way to test this (either code and/or encrypted streams) may be specific to a particular DRM system.

**Comment by cta-source** [comment]
From dpctf@standards.cta.tech email _"Top-down look at cDRM"_ on 2025-02-12:

Summarizing what I’ve read in the DPCTF slides today, this seems to be the hierarchy of the commercial DRM (“cDRM”) project.

1.	License Server (required for Content Encoding steps)
a.	Who provides? 
2.	Content
a.	Content encoding & packaging workflow update
b.	Content generation using existing mezzanine content
3.	Tests 
a.	Per [DPCTF repo Issue #118](https://github.com/cta-wave/device-playback-task-force/issues/118), there's a number of different groups of tests.
i.	Repetitions of existing tests but with encrypted content (assume this is the minimum)
ii.	General DRM features
iii.	DRM system specific features
b.	Extend Fraunhofer HTML+JS test templates 
i.	…need extending to request information from a DRM license server and pass the information returned to the EME decryption mechanism.
4.	Test Runner
a.	What?
5.	Observation Framework
a.	Map to “Tests” in Item 3 above. 
i.	Repetitions of existing tests (existing)
ii.	General DRM features (new)
iii.	DRM system specific features (new)

**Comment by jpiesing** [comment]
This is what I wrote during the meeting on February 18th.

See https://github.com/cta-wave/device-playback-task-force/issues/118​

Steps​
- Define basic functional requirements for what we're testing​
- Define our expectations from supplier / partner​
-- Non-technical requirements, agreements, etc incl. who can use this?​
-- Annual fee may be desirable anyway​
-- Who can't use this may require legal advice
- Select supplier / partner (EasyDRM, others)​
- Agree list of streams to be encrypted​
- Agree interface between Romain's scripts & partner license server​
- Update scripts to generate encrypted streams, generate them & validate them​
- Update existing encrypted content tests to request a DRM license​
- Integrate​
- Validate​

**Comment by cta-source** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MBergman/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MBergman/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


Category   / Task | Output
-- | --
**Requirements Planning** |  
Define basic functional   requirements for what we're testing​ | Functional requirements document
Agree list of streams to be   encrypted​ | List of streams
Define our expectations from   supplier / partner​ |  
-- Non-technical requirements,   agreements, etc incl. who can use this?​ | Summary from Counsel (later: contracts if needed)
-- Determine if a fee (annual,   e.g.) should be required of users. | Preliminary fee structure, if any
**Partner Selection** |  
Develop   request-for-information or other documentation on partnerships for use in   supplier selection | RFI or other communications document
Select supplier / partner   (EasyDRM, others)​ | Decision on partner
**Partnership Activities** |  
Agree interface between   Romain's scripts & partner license server​ | Summary of interface
**Content** |  
Content encoding &   packaging workflow update | Updated scripts
Content generation and   validation using existing mezzanine content | New content
**Test Development** |  
Per DPCTF repo Issue   #118, there's a number of different groups of tests. |  
-- Repetitions of existing tests   but with encrypted content (assume this is the minimum) |  
-- General DRM features | New tests, "general DRM"
-- DRM system specific features | New tests, "DRM system specific"
Extend Fraunhofer HTML+JS test   templates for license fetch, referral to EME | Updated test templates
**Test Runner Development** |  
(tbd -- needs expansion) |  
**Observation Framework Development** |  
Map to tests as per Test   Development section |  
-- Repetitions of existing tests   (existing) |  
-- General DRM features (new) | Updated OF for "general DRM"
-- DRM system specific features   (new) | Update OF for "DRM system specific"
**Deploy Development** |  
Update instructions to include   any requirement for DRM testing |  
**Integration** |  
Integrate​ |  
Validate​ |  
Validate Deploy and   instructions |  



</body>

</html>

This chart merges the above two comments (i.e., the ones immediately above this comment, by @jpiesing and @cta-source, both comments posted 2025-02-18). Added an "Output" column and made guesses as to the logical work product of most tasks. Liberal edits made en route to the result.

**Comment by jpiesing** [comment]
I've reviewed https://github.com/cta-wave/device-playback-task-force/issues/118#issuecomment-2666402513 and only have a few comments.
* I can't think of any "Test Runner Development" and we should define the first phase such that any test or feature that needs it is automatically deferred. The closest we may get to that is tests that have a time of day dependency, e.g. licenses that expire at a particular date & time. Something (but not the test runner itself) would need to download new content periodically.
* Further we should define the first phase such that any test or feature that needs changes to the OF is automatically deferred.
* In the February 12th DPCTF meeting, one of the discussion points was about the interface between Romain's scripts and the partner license server. Does this need to be automated or could it just be manual? Manual would be simpler but would limit us to test content that effectively had an infinite non-expiring license - if that's possible.

**The Absolute Minimum**

We have 3 templates for encrypted content.
* playback-of-encrypted-content. This tests playing a 30s stream from beginning to end.
* restricted-splicing-of-encrypted-content
* sequential-playback-of-encrypted-and-non-encrypted-baseline-content

The **absolute minimum** therefore is to run each of these 3 with video (perhaps just AVC or just HEVC?), each with audio (but can we agree on one codec?), and each once with each DRM and once with all 3 DRMs. For the individual DRM tests, for PlayReady and Widevine, we should use CENC and but obviously have to use CBCS 1:9 for the stream encrypted with all 3 DRMs.

I think this would mean the following ...

1. playback-of-encrypted-content - video
2. restricted-splicing-of-encrypted-content - video
3. sequential-playback-of-encrypted-and-non-encrypted-baseline-content - video
4. playback-of-encrypted-content - audio
5. restricted-splicing-of-encrypted-content - audio
6. sequential-playback-of-encrypted-and-non-encrypted-baseline-content - audio

That would then be multiplied by 4 for each DRM individually and multi-DRM giving a total of 24.
If we can't agree on a single video codec and a single audio codec then doubling it gets to 48.

I think this would mean the following encrypted streams.
- Video stream t1 in 4 versions - PlayReady/CENC, Widevine/CENC, FairPlay/CBCS and all 3 (CBCS).
- Video streams splice_main and splice_ad each in 4 versions as above
- Audio streams equivalent to at5, at15 and at16 (which are all MPEG so Dolby versions may be needed) also in 4 versions.

**Slightly more than the minimum**

The most obvious addition is a stream with both video and audio, e.g.  clone of our current test "regular-playback-of-a-cmaf-presentation" but with EME copied over from playback-of-encrypted-content (as above).

Much more than this risks trying to run before we can walk.

**Comment by mbergman42** [comment]
*   I can't think of any "Test Runner Development" and we should define the first phase such that any test or feature that needs it is automatically deferred. The closest we may get to that is tests that have a time of day dependency, e.g. licenses that expire at a particular date & time. Something (but not the test runner itself) would need to download new content periodically.
Agreed. I’ve captured this point in the spreadsheet version of this table.

  *   Further we should define the first phase such that any test or feature that needs changes to the OF is automatically deferred.
Sorry, which “first phase” is this? We’ve spoken of doing some work in 2025 without trying to do a full project, but I’ve been looking for an end-to-end project plan that ensures we don’t get stuck with development in 2025 that cannot be used in the final version.
So are you speaking of “first phase is a partial development stage, en route to a usable test suite after further phases”, or “first phase leads to a full usable tool but may miss some desired features”?

  *   In the February 12th DPCTF meeting, one of the discussion points was about the interface between Romain's scripts and the partner license server. Does this need to be automated or could it just be manual? Manual would be simpler but would limit us to test content that effectively had an infinite non-expiring license - if that's possible.
I’ll leave this to others…

Mike Bergman
Consumer Technology Association, producer of CES®
m: +1(609) 865-4402

**Comment by jpiesing** [comment]
> Sorry, which “first phase” is this? We’ve spoken of doing some work in 2025 without trying to do a full project, but I’ve been looking for an end-to-end project plan that ensures we don’t get stuck with development in 2025 that cannot be used in the final version.
> So are you speaking of “first phase is a partial development stage, en route to a usable test suite after further phases”, or “first phase leads to a full usable tool but may miss some desired features”?

The latter.
This is a large and complex problem and my assumption is that we do not have the budget to pay for all of it in 2025 or the bandwidth to review and absorb a complete set of results in 2025.
Hence start with a subset of tests in 2025 and add more later.
Going back to the very first comment in this issue, a minimum useful subset is "Repetitions of existing tests but with encrypted content".
Of course nothing is black and white & there may be items from the list in that issue that can be cherry picked without a big impact on cost, schedule or resources.

**Comment by mbergman42** [comment]
OK. I’m updating the plan to include:

Requirements Planning
Define basic functional requirements for what we're testing​
Phase 1 (basic usable system based on "repetitions of existing tests")
Phase 2 (adding some of: general DRM features and DRM specific features)
Phase 3 and beyond (finish features from Phase 2, other enhancements)


Mike Bergman
Consumer Technology Association, producer of CES®
m: +1(609) 865-4402

**Comment by cta-source** [comment]
I've incorporated the above discussion in the spreadsheet version of the project outline, which you can see in table form in a prior comment (not visible in email, but visible as a comment on 2/18 on the browser view of this issue).
